allennlp/modules/elmo.py

```diff
@@ -681,3 +685,148 @@ def create_cached_cnn_embeddings(self, tokens: List[str]) -> None:
+    def _load_softmax_weights(
+            softmax_weight_file: str) -> torch.nn.Linear:
+
+        with h5py.File(cached_path(softmax_weight_file), 'r') as fin:
+            W = numpy.array(fin['softmax']['W'])
+            b = numpy.array(fin['softmax']['b'])
+
+        vocab_size = W.shape[0]
+        hidden_size = W.shape[1]
+        fc_layer = torch.nn.Linear(hidden_size, vocab_size)
+
+        fc_layer.weight.data.copy_(
+            torch.FloatTensor(W))
+        fc_layer.bias.data.copy_(
+            torch.FloatTensor(b))
+
+        return fc_layer
+
+    def _chunked_log_probs(self,
+                           activation: torch.Tensor) -> torch.Tensor:
+        """
+        Do the softmax in chunks so the gpu ram doesn't explode.
+
+        Parameters
+        ----------
+        activation : ``torch.Tensor``, required
+            Results of the ELMo embedder.
+
+        Returns
+        -------
+        ``torch.Tensor`` with all of the log probabilities.
+        """
+        all_logprobs = []
+        num_chunks = (activation.size(0) - 1) // self.chunk_size + 1
+        for activation_chunk in torch.chunk(activation, num_chunks, dim=0):
+            chunk_size, sequence_length = activation_chunk.size()[:2]
+
+            time_indexer = (torch.arange(0, chunk_size * sequence_length)
+                            % sequence_length)
+
+            batch_indexer = (torch.arange(0, chunk_size * sequence_length)
+                             / sequence_length)
+
+            all_logprobs.append(
+                F.log_softmax(self.fc_layer(activation_chunk), 2)[
+                batch_indexer, time_indexer, :].view(
+                    *[chunk_size, sequence_length, -1]))
+
+        return torch.cat(all_logprobs, 0)
+
+    def forward(self,
+                bilm_inputs: Dict[str, Union[torch.Tensor, List[torch.Tensor]]],
+                aggregation_function: str = None) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Passes the ELMo embeddings to a softmax layer and calculates
+        the log probabilities.
+
+        Parameters
+        ----------
+        bilm_inputs : ``Dict[str, Union[torch.Tensor, List[torch.Tensor]]]``, required
+            The output from `_ElmoBiLm`.
+        word_inputs : ``torch.Tensor``, required
+            The output from `batch_to_ids` when `
+        aggregation_function : ``str``, optional
+            An aggregation function that aggregates the forward and backward
+            tensors. Default is `None` where no aggregation is done.
+
+        Returns
+        -------
+        A tuple of `log_probs` and `mask_without_bos_eos`.
+        """
+
+        activations = bilm_inputs['activations']
+        mask = bilm_inputs['mask']
+
+        final_activations = activations[-1]  # (batch, BOS + L + EOS, [fwd, bwd])
+
+        # get rid of predicting, or conditioning on EOS for the forward
+        # activation, and BOS for the reverse activation.
+        mask_without_bos_eos = mask[:, 2:]
```

Comment:
Slicing like this won't work at all, you need to take into account the mask and padding.  Use this function to remove the sentence boundaries from both the mask and the `final_activations`: https://github.com/allenai/allennlp/blob/master/allennlp/nn/util.py#L1193